src/bobtail/dag.cpp

```diff
@@ -117,6 +86,89 @@ void CBobtailDag::UpdateCompatibility(const int16_t &new_id, const std::set<int1
     }
 }
 
+void CBobtailDag::UpdateDagScore()
+{
+    // keep track of what has been seen
+    std::map<CDagNode*, uint64_t> seen;
+    // build out the dag by level, a nodes level is determined by
+    // its shortest path to a base
+    std::vector<std::set<CDagNode*> > leveled_dag;
+    // first find the bases
+    leveled_dag.emplace_back();
+    bool do_another_level = false;
+    // there is always at least 1 node in a dag
+    for (auto &node : _dag)
+    {
+        if (node->IsBase())
+        {
+            leveled_dag[0].emplace(node);
+            seen.emplace(node, 0);
+            if (node->descendants.empty() == false)
+            {
+                do_another_level = true;
+            }
+        }
+    }
+    while (do_another_level == true)
+    {
+        size_t search_index = leveled_dag.size() - 1;
+        // add another level
+        leveled_dag.emplace_back();
+        do_another_level = false;
+        for (auto &node : leveled_dag[search_index])
+        {
+            for (auto &desc : node->descendants)
+            {
+                if (seen.count(desc) == 0)
+                {
+                    seen.emplace(desc, 0);
+                    leveled_dag.back().emplace(desc);
+                    if (desc->IsBase())
+                    {
+                        do_another_level = true;
+                    }
+                }
+            }
+        }
+    }
+    // calculate the score
+    uint16_t total_score = 0;
+    // set all of the tips to a score of 1
```

Comment:
Not sure why this loop is necessary, i.e. why do the tips need to be treated differently than the other nodes?